### PR TITLE
Add IRCv3.2 tag compatibility

### PR DESCRIPTION
**Pull request by ChickenNuggers**
The [IRCv3](http://ircv3.net/specs/core/message-tags-3.2.html) spec defines a way to add in tags, which are somewhat like a command metadata; this change is supposed to add compatibility with them.

@daurnimator since you had questions about the code length, maybe you could see something to optimize?

### CODE REVIEW

File: init.lua [hunk 1]
```diff
@@ -111,9 +111,67 @@ end
 
 
 --- Receiving ---
+-- IRCv3.2 tags
+local function parse_tags(tag_message)
+	local tags = {}
+	local is_on_name = true
+	local cur_name = ""
+	local escaping = false
+	local escapers = {
```
Comment (daurnimator) [inline init.lua]: hoist this out.

File: init.lua [hunk 2]
```diff
@@ -111,9 +111,67 @@ end
...
+		["s"] = " ",
+		["r"] = "\r",
+		["n"] = "\n",
+		[";"] = ";"
+	}
+	local charbuf = ""
```
Comment (daurnimator) [inline init.lua]: Never append to a string in a loop.
Build up a table and use table.concat at the end.

File: init.lua [hunk 3]
```diff
@@ -111,9 +111,67 @@ end
...
+			charbuf = ""
+		elseif char == ";" then
+			if not is_on_name then
+				is_on_name = true
+				tags[cur_name] = charbuf
+				cur_name = ""
+				charbuf = ""
+			else
+				tags[charbuf] = true
+				charbuf = ""
```
Comment (daurnimator) [inline init.lua]: missing `is_on_name = false`?

File: init.lua [hunk 4]
```diff
@@ -111,9 +111,68 @@ end
...
+	end
+	return tags
+end
+
 -- Main --
 -- http://calebdelnay.com/blog/2010/11/parsing-the-irc-message-format-as-a-client
-local function parse_message(message)
+local function parse_message(message_tagged)
+	-- Tags
+	local tagged, tags, message
```
Comment (legospacy) [inline init.lua]: Is the variable `tagged` assigned to or used anywhere? I don't see anywhere it's actually used...

File: init.lua [hunk 5]
```diff
@@ -111,9 +111,68 @@ end
...
+	if message:sub(1, 1) == "@" then
```
Comment (legospacy) [inline init.lua]: Should `message:sub(1, 1)` be `message_tagged:sub(1, 1)`?

**Comment by legospacy** [comment]
Thanks for the code!

Looking at `http://ircv3.net/specs/core/message-tags-3.2.html` and trying out the code: 

Does the tag part of a message always end with a semicolon? From what I can tell, the last tag doesn't have terminating semicolon, and when I run the code on tags without that semicolon it drops the last tag.

**Comment by legospacy** [comment]
Oh, also, that page also says the escape character for semicolons is actually a normal colon, (`; = \:`), while this code seems to have it as a semicolon.

**Comment by ChickenNuggers** [comment]
I'll go double check that in a bit.

On Tue, Jan 26, 2016 at 6:15 PM legospacy <notifications@github.com> wrote:

> Oh, also, that page also says the escape character for semicolons is
> actually a normal colon, (; = \:), while this code seems to have it as a
> semicolon.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/legospacy/lua-irc-engine/pull/11#issuecomment-175307674>
> .
>
> Augusta Public Schools USD 402- Student Email

-- 

Charles Heywood <vandor2012@gmail.com>
Programmer :: Scripter :: Designer :: Administrator
https://github.com/ChickenNuggers

**Comment by ChickenNuggers** [comment]
> Does the tag part of a message always end with a semicolon? From what I can tell, the last tag doesn't have terminating semicolon, and when I run the code on tags without that semicolon it drops the last tag.
That might be my fault also; I'll fix that as well.

**Comment by ChickenNuggers** [comment]
@legospacy I double checked, I don't think that there's a place where : is an escaper. Also, I made sure that the message parsing function returns `tags` and fixed the occurance of `message` => `tagged_message`. Anything else?

**Comment by legospacy** [comment]
http://ircv3.net/specs/core/message-tags-3.2.html, under "Escaping values":
>Reason: more common URL-escaping eats more space, while IRC message’s length is very limited. Also having semicolon as \: makes it easy to split the <tags> string by ; first.

**Comment by legospacy** [comment]
Apart from dropping the last tag if there's no trailing semicolon (eg. `@aaa=bbb;ccc;example.com/ddd=eee` seems to only return `{aaa = bbb, ccc = true}`, while putting a semicolon on the end returns all three correctly), I think it's all good!

**Comment by ChickenNuggers** [comment]
The semicolon isn't required; I'll fix that now.

On Tue, Jan 26, 2016 at 9:35 PM legospacy <notifications@github.com> wrote:

> Apart from dropping the last tag if there's no trailing semicolon (eg.
> @aaa=bbb;ccc;example.com/ddd=eee seems to only return {aaa = bbb, ccc =
> true}, while putting a semicolon on the end returns all three correctly),
> I think it's all good!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/legospacy/lua-irc-engine/pull/11#issuecomment-175372862>
> .
>
> Augusta Public Schools USD 402- Student Email

-- 

Charles Heywood <vandor2012@gmail.com>
Programmer :: Scripter :: Designer :: Administrator
https://github.com/ChickenNuggers

**Comment by legospacy** [comment]
Awesome! Sorry to keep asking about things, but I'm still not entirely convinced that escaping semicolons is done with `\;`, the page seems to suggest `\:'. The table under "Escaping values" does have `;` being escaped as `\:`. What do you think?

**Comment by legospacy** [comment]
Ah well, it's a little thing. Thanks again!

**Pull request merged**